File: Root_to_leaf.java
//Time-O(n)
//Space-O(n)-Recursive stack in consideration
//Had the iterative approach using string but found this better 
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int sum;
    public int sumNumbers(TreeNode root) {
        sum=0;
        recurr(root,0);
        return sum;
    }
    
    public void recurr(TreeNode root, int curr)
    {
        if(root==null)
        {
            return;
        }
        curr=curr*10+root.val;
        if(root.left==null && root.right==null)
        {
            sum=sum+curr;
            return;
        }
        recurr(root.left,curr);
        recurr(root.right,curr);
        
    }
}

File: root_to_leaf.py
"""
TC:O(n)
SC:O(h), h being height of tree
"""
class Solution:
    def sumNumbers(self, root: TreeNode):
        def preorder(r, curr_number):
            nonlocal root_to_leaf
            if r:
                curr_number = curr_number * 10 + r.val
                # if it's a leaf, update root-to-leaf sum
                if not (r.left or r.right):
                    root_to_leaf += curr_number
                preorder(r.left, curr_number)
                preorder(r.right, curr_number)
        root_to_leaf = 0
        preorder(root, 0)
        return root_to_leaf
